fix(sink): resolve schema mismatch due to row_id column

nanderstabel opened the pull request:
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Fixes #4303

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

When writing to a mysql table, an schema mismatch error occurred due to writing the `_row_id` column along with the intended columns. In the initial version of this PR I "fixed" the issue by simply omitting columns suffixed with "_row_id" when writing to MySQL. However, this solution was lacking some nuance. 

Hidden columns should not simply be omitted in SinkExecutor. In the current version all columns, including hidden columns are written to MySQL.

This PR:
- Adds `fn prepare()` to `SinkImpl` in order to make sure (in case of MySQLSink) that a table with correct schema (including hidden columns)  before data is written to the external sink.
- Makes sure sink is properly dropped in stream manager (closes: #4302).
- Tests

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

neverchanje commented on e2e_test/sink/t3.slt:
```diff
@@ -0,0 +1,23 @@
+
+statement ok
+INSERT INTO t3 VALUES (1, 2, 3), (1, 2, 3), (2, 2, 3), (3, 2, 3), (5, 2, 3), (8, 2, 3), (13, 2, 3), (21, 2, 3);
+
```
```suggestion
statement ok
flush
```

Better to ensure the inserts are flushed

neverchanje reviewed (commented):
Rest lgtm

tabVersion commented:
Please hold for a while

skyzh commented on src/connector/src/sink/mysql.rs:
```diff
@@ -163,7 +183,15 @@ async fn write_to_mysql<'a>(
+            .zip_eq(schema.names().iter())
+            .filter_map(|(c, n)| {
+                // Exclude row id column.
+                if !n.ends_with("_row_id") {
```
I'm afraid this is too ad-hoc for sink. It would be better to add a projection before sink in frontend plan to remove this column if you don't want it.

skyzh commented:
After a second thought, it can't be correct if we need to manually remove a column like this. For agg MV, the pk will be the group key. For join MV, the pk will be join key + original pk. Removing row_id might make sense for table source, but for other sources there might not be a column called "row_id" and might have something more to remove.

The logic needs to be handled in frontend instead of in sink. Need revisit the code instead of merging this PR.

tabVersion commented:
agree with @skyzh , sink have no following executors, its column (or visibility) should be aligned to external tables

nanderstabel commented:
@skyzh @tabVersion I agree as well, so I've updated the code. Main difference is that no columns (including hidden ones) are omitted anymore. I also added `fn prepare` method to `SinkImpl` which ensures a table with correct schema is created in MySQL. Perhaps at a later stadium an option can be added to Sink's properties so that the user can decide whether not to include hidden columns.

(I've updated the description of this PR as well)

tabVersion commented on src/connector/src/sink/mysql.rs:
```diff
@@ -125,8 +149,57 @@ impl TryFrom<Datum> for MySQLValue {
 }
 
+#[allow(clippy::upper_case_acronyms)]
+#[derive(strum_macros::Display)]
```
what does `strum_macros::Display` do

nanderstabel replied:
proc_macro Display converts enum variants to strings, i.e. `MySQLDataType::BOOL` will be printed as `"BOOL"`. In this case I used it for the schema in the create table statement.

tabVersion commented on src/connector/src/sink/mysql.rs:
```diff
@@ -125,8 +149,57 @@ impl TryFrom<Datum> for MySQLValue {
+
 #[async_trait]
 impl Sink for MySQLSink {
+    async fn prepare(&mut self, schema: &Schema) -> Result<()> {
```
I really doubt whether we should do that, cc @neverchanje .

nanderstabel replied:
Perhaps adding this prepare function is not the best way. Another solution is to move the "prepare" implementation to `MySQLSink::new`. This might make more sense and won't require other Sinks to implement a prepare function. The only change this will require is that `schema` needs to be added to `fn build_sink` and `SinkImpl::new` signature. Wdyt @tabVersion ?

tabVersion commented on src/stream/src/executor/sink.rs:
```diff
@@ -151,27 +156,56 @@ impl<S: StateStore> Executor for SinkExecutor<S> {
 
-        // Mock `child`
-        let _mock = MockSource::with_messages(Schema::default(), PkIndices::new(), vec![]);
+        let mut properties = HashMap::new();
```
you can try `maplit::hashmap` here

tabVersion reviewed (commented):
as commented before, in my view, hidden columns are not visible to users and external components. A projection may need here.

skyzh commented:
hidden columns (generally stream pk) is crucial to correctness. We should always include them. One example:


If users create a MV which contains exactly one visible column A, with two rows of the same value 1. Without hidden column of stream key, we can never distinguish them.

skyzh commented:
... which leads to inconsistency between RisingWave's query result and sink's stored result during updates and deletes.

skyzh commented:
... and thus I believe this impl is correct. No need to use projections to remove hidden columns cc @tabVersion

skyzh reviewed (approved):
LGTM from my side. May wait for others' comments.

neverchanje commented:
> hidden columns (generally stream pk) is crucial to correctness. We should always include them. One example:
> 
> If users create a MV which contains exactly one visible column A, with two rows of the same value 1. Without hidden column of stream key, we can never distinguish them.

Good point. I agree that we should include `_row_id` to the sink table in case it has not declared any primary key column. Then the two records will be seen as undistinguishable from our side. It should have a `_row_id` to allow us uniquely update its value.

skyzh commented:
> it has not declared any primary key column

If MySQL has declared pk column, we should do an extra shuffle on our side, and we'll need to specify new pk in `create_sink` statement. May discuss later.

tabVersion commented:
> If MySQL has declared pk column, we should do an extra shuffle on our side, and we'll need to specify new pk in create_sink statement. May discuss later.

Tao is on a new design. One thing for sure: we will have pk on sink
